server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.telemetry.TelemetryProvider;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+public class SearchTransportAPMMetrics {
+    private final LongHistogram actionLatencies;
```

Comment:
Question: we are going to store latencies for different phases using this same variable, but with different attributes? Intuitively, and maybe naively, i would have expected one variable per type of latency.

Reply:
They way I think about it (and I could be wrong), is that each of the possible combinations  of the "static" name and the "dynamic" attributes can be considered a separate time-series.  The attributes are useful when performing aggregations.  I was thinking it could be the case that we'd want to group various phases together in some of our reporting (with the option of separating them when drilling down), but I could be mistaken 🙂  WDYT?
 
 The [proposal](https://docs.google.com/document/d/1jKxuaZi7QAMIRD_Eq3nonkYswVlQXW5TWllJEicoOtM/edit) mentions:
- When there are different "flavors" of a metric (i.e. s3, azure, etc) use an attribute rather than inserting it in the metric name.
- Rule of thumb: you should be able to do aggregations (e.g. sum, avg) across a dimension of a given metric (without the need to aggregate over different metric names); on the other hand, any aggregation across any dimension of a given metric should be meaningful.

Reply:
I find this hard to follow. Do you have an example of pros and cons of the two approaches (single metric vs multiple metrics)?

Reply:
I was thinking we'd want to monitor/chart transport action latencies in general (or at least a subset of them that we care about). 

In any of these latency metrics, the value at any given time matters little. What  matters (IMO) is trends: changes in metric values in aggregate over time.  If we see transport actions latencies increase dramatically for a given node, we know there's a problem.  If we see transport actions that are more expensive across the board for a given cluster/cloud-provider/region/version/etc we know there's a problem. We can gain that insight more easily if we have a single metric that we monitor, rather than manually defining the chart/alert as a collection of metrics which we must remember to update if the code changes.  

I don't think we loose anything by labeling the transport actions as attributes on the metric, as we can always filter for specific actions to "drill down".

At least, that's my understanding. Much of my metrics understanding was learned from using other time-series databases, so  I'm happy to be persuaded otherwise though if you're thinking of something I'm missing!

Metrics can be thought of as an API, and APIs are hard to change. I think we should have a conversation about this to ensure we get this right.

Reply:
As far as I understand there isn't a huge difference between single metric and multiple metrics. I am fine with using one with attributes.